### PR TITLE
Playwright test framework improvement

### PR DESCRIPTION
**Pull request by aweng98**
Major PW test framework overhaul.

- Additional enums re-org.
- Removed duplicate functions.
- DSM page files renamed and re-grouped.
- For others, see inline comments.

`/dsm/pages` new hier.

Screenshot 2024-03-04 at 1 55 26 PM

### CODE REVIEW

File: playwright-e2e/dsm/component/dropdown.ts [hunk 1]
```diff
@@ -40,7 +38,7 @@ export default class Dropdown {
     return this.toLocator().locator('ul.dropdown-menu').locator('a').getByText(value, {exact: true});
   }
 
-  async getDisplayedOptions<T extends MiscellaneousEnum | SamplesNavEnum | StudyNavEnum | StudyEnum>(): Promise<T[]> {
+  async getDisplayedOptions<T extends Miscellaneous | Samples | Study | StudyName>(): Promise<T[]> {
```
Comment (aweng98) [inline playwright-e2e/dsm/component/dropdown.ts:41]: Consolidated different navigation enums to one file `navigation.ts`. Renamed enums to make them shorter.

File: playwright-e2e/dsm/component/modal.ts [hunk 2]
```diff
@@ -5,27 +5,28 @@ import Radiobutton from 'dss/component/radiobutton';
 import { waitForNoSpinner } from 'utils/test-utils';
 
 export default class Modal {
-  private readonly rootSelector: Locator;
+  private readonly root: Locator;
 
-  constructor(private readonly page: Page) {
-    this.rootSelector = this.page.locator('.modal-dialog').locator('visible=true');
+  constructor(private readonly page: Page, root?: Locator) {
```
Comment (aweng98) [inline playwright-e2e/dsm/component/modal.ts:10]: adding root (parent) locator for exact searching.

File: playwright-e2e/dsm/component/sm-id.ts [hunk 3]
```diff
@@ -11,8 +11,8 @@ interface InputData {
 export default class SMID {
   private readonly modalComponent: Modal;
 
-  constructor(private readonly page: Page, private readonly tissueIndex: number) {
-    this.modalComponent = new Modal(this.page);
+  constructor(private readonly page: Page, private readonly root: Locator) {
+    this.modalComponent = new Modal(page, root);
```
Comment (aweng98) [inline playwright-e2e/dsm/component/sm-id.ts:15]: Instead using an index number, new param `root` locator is used to tie tissue and SM-ID dialog together.

File: playwright-e2e/dsm/component/tabs/enums/onc-history-input-columns-enum.ts [hunk 4]
```diff
@@ -1,27 +1,3 @@
-export enum OncHistoryInputColumnsEnum {
```
Comment (aweng98) [inline playwright-e2e/dsm/component/tabs/enums/onc-history-input-columns-enum.ts:1]: dup of existing enums in `enums.ts`

File: playwright-e2e/dsm/component/modal.ts [hunk 5]
```diff
@@ -5,27 +5,28 @@ import Radiobutton from 'dss/component/radiobutton';
...
+    const loc = root ? root : this.page.locator('app-root app-all-studies');
+    this.root = loc.locator('.modal-dialog').locator('visible=true');
   }
 
-  public toLocator(): Locator {
-    return this.rootSelector;
+  public get toLocator(): Locator {
```
Comment (aweng98) [inline playwright-e2e/dsm/component/modal.ts:15]: Make it `get` accessor. Use it without `()`.

File: playwright-e2e/dsm/pages/dsm-page-base.ts [hunk 6]
```diff
@@ -23,8 +26,40 @@ export default abstract class DsmPageBase {
...
+      await this.page.waitForTimeout(200);
+    }
+    if (value) {
+      const resPromise = waitForResponse(this.page, {uri: '/patch'});
+      await input.fillSimple(value);
+      await resPromise;
+    }
+  }
+
+  public async backToParticipant(): Promise<void> {
```
Comment (aweng98) [inline playwright-e2e/dsm/pages/dsm-page-base.ts:51]: common func

File: playwright-e2e/dsm/pages/dsm-page-base.ts [hunk 7]
```diff
@@ -23,8 +26,40 @@ export default abstract class DsmPageBase {
...
+    await this.page.getByText(/back to (Participant|previous) page/i).click();
+    await expect(this.page.locator('//app-participant-page/h1')).toHaveText('Participant Page');
+    await waitForNoSpinner(this.page);
+  }
+
+  public async backToParticipantList(): Promise<void> {
```
Comment (aweng98) [inline playwright-e2e/dsm/pages/dsm-page-base.ts:57]: common func

File: playwright-e2e/tests/dsm/onc-history/cmi-clinical-sequencing-order.spec.ts [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+import TissueRequestPage from 'dsm/pages/tablist-pages/tissue-request-page';
+import crypto from 'crypto';
+
+/**
+ * Create an Onc History with accession number, gender, etc.
+ * Create a new tissue with date sent to GP and number of SM-IDS
+ * Mark it as received through API "/ddp/ClinicalKits"
+ */
+
+test.describe.skip('Create Onc History', () => {
```
Comment (aweng98) [inline playwright-e2e/tests/dsm/onc-history/cmi-clinical-sequencing-order.spec.ts:22]: New test PEPPER-1108: waiting for bug fixes in Tissue Request page.

File: playwright-e2e/authentication/auth-dsm.ts [hunk 9]
```diff
@@ -45,7 +45,7 @@ export async function login(page: Page, opts: { email?: string; password?: strin
     return;
   }
   // Retry login if login window is still present (POST /auth0 fails intermittenly)
-  console.log(`Retry log in`);
+  console.log(`      Retry log in`);
```
Comment (Rinchan20) [inline playwright-e2e/authentication/auth-dsm.ts]: should this be logInfo?

Comment (aweng98) [reply playwright-e2e/authentication/auth-dsm.ts]: used to help me to see retry works or not. I don't need it anymore, will remove.

File: playwright-e2e/dsm/pages/tablist/sequencing-order-tab.ts [hunk 10]
```diff
@@ -1,23 +1,25 @@
...
   }
 
   /* Actions */
 
   public async waitForReady(): Promise<void> {
-    const sequencingTab = this.page.locator(`//tab[@heading='Sequencing Order']`);
-    await sequencingTab.scrollIntoViewIfNeeded();
-    await expect(sequencingTab, 'The Sequencing Tab is not currently active').toHaveClass(/active/);
+    await this.toLocator.scrollIntoViewIfNeeded();
+    await expect(this.toLocator, 'The Sequencing Tab is not currently active').toHaveClass(/active/);
```
Comment (aweng98) [inline playwright-e2e/dsm/pages/tablist/sequencing-order-tab.ts:18]: `TabBase.ts` takes care constructing the tab xpath automatically. Use `toLocator` to find elements inside the opened tab.

File: playwright-e2e/dsm/component/tablist.ts [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+    return clas ? clas.includes('active') : false;
+  }
+
+  public async click<T extends object>(): Promise<T> {
+    await expect(this.toLocator, `Tab "${this.tab}" is not enabled`).toBeEnabled();
+    await this.checkAndClick();
+    return (this.tabs as Map<string, object>).get(this.tab) as T;
+  }
+
+  protected async HasContactInformationData(): Promise<boolean> {
```
Comment (Rinchan20) [inline playwright-e2e/dsm/component/tablist.ts]: small thing but h in has should probably be lower case

Comment (aweng98) [reply playwright-e2e/dsm/component/tablist.ts]: Good catch 👍🏻

**Review by Rinchan20 — APPROVED** [review]

**Pull request merged**